pkg/joinserver/http_interop.go

```diff
@@ -16,17 +16,142 @@ package joinserver
 
 import (
 	"context"
+	"time"
 
+	"go.thethings.network/lorawan-stack/pkg/encoding/lorawan"
 	"go.thethings.network/lorawan-stack/pkg/errors"
 	"go.thethings.network/lorawan-stack/pkg/interop"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
 )
 
-type interopServer struct {
-	JS *JoinServer
+type interopHandler interface {
+	HandleJoin(context.Context, *ttnpb.JoinRequest) (*ttnpb.JoinResponse, error)
+	GetHomeNetID(context.Context, types.EUI64, types.EUI64) (*types.NetID, error)
+	GetAppSKey(context.Context, *ttnpb.SessionKeyRequest) (*ttnpb.AppSKeyResponse, error)
 }
 
-var errInteropNotImplemented = errors.DefineUnimplemented("interop_not_implemented", "interop endpoint not implemented")
+type interopServer struct {
+	JS interopHandler
+}
 
 func (srv interopServer) JoinRequest(ctx context.Context, req *interop.JoinReq) (*interop.JoinAns, error) {
-	return nil, errInteropNotImplemented
+	ctx = log.NewContextWithField(ctx, "namespace", "joinserver/interop")
+
+	var cfList *ttnpb.CFList
+	if len(req.CFList) > 0 {
+		cfList = new(ttnpb.CFList)
+		if err := lorawan.UnmarshalCFList(req.CFList, cfList); err != nil {
+			return nil, interop.ErrMalformedMessage.WithCause(err)
+		}
+	}
+	var dlSettings ttnpb.DLSettings
+	if err := lorawan.UnmarshalDLSettings(req.DLSettings, &dlSettings); err != nil {
+		return nil, interop.ErrMalformedMessage.WithCause(err)
+	}
+
+	res, err := srv.JS.HandleJoin(ctx, &ttnpb.JoinRequest{
```

Comment:
The `JoinRequest` is validated before passed to `HandleJoin` via gRPC, shall we also validate it here(since the interceptor is not called)?

Reply:
Yes, good catch